// @microsoft/teamsfx-core/build/question/scaffold/constructNode.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructNode = void 0;
const localizeUtils_1 = require("../../common/localizeUtils");
const commonNodes_1 = require("./commonNodes");
const CapabilityOptions_1 = require("./vsc/CapabilityOptions");
function constructNode(jsonContent) {
    const jsonObject = JSON.parse(jsonContent);
    if (jsonObject.node === "llmServiceNode") {
        return (0, commonNodes_1.llmServiceNode)(jsonObject.condition);
    }
    else if (jsonObject.node === "apiSpecNode") {
        return (0, commonNodes_1.apiSpecNode)(jsonObject.condition);
    }
    else if (jsonObject.node === "azureOpenAINode") {
        return (0, commonNodes_1.azureOpenAINode)(jsonObject.condition);
    }
    else if (jsonObject.node === "foundryNode") {
        return (0, commonNodes_1.foundryNode)(jsonObject.condition);
    }
    const data = {
        type: jsonObject.data.type,
        name: jsonObject.data.name,
        title: (0, localizeUtils_1.getLocalizedString)(jsonObject.data.title),
        placeholder: (0, localizeUtils_1.getLocalizedString)(jsonObject.data.placeholder),
        onDidSelection: CapabilityOptions_1.setTemplateName,
        staticOptions: [],
    };
    for (const option of jsonObject.data.options) {
        data.staticOptions.push({
            id: option.id,
            label: (0, localizeUtils_1.getLocalizedString)(option.label),
            detail: (0, localizeUtils_1.getLocalizedString)(option.detail),
            data: option.data,
        });
    }
    const node = {
        data: data,
    };
    if (jsonObject.condition) {
        node.condition = jsonObject.condition;
    }
    for (const option of jsonObject.children) {
        const childNode = constructNode(JSON.stringify(option));
        node.children = node.children || [];
        node.children.push(childNode);
    }
    return node;
}
exports.constructNode = constructNode;
