// @microsoft/teamsfx-core/build/core/FxCore.declarativeAgent.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.FxCoreDeclarativeAgentPart = void 0;
const tslib_1 = require("tslib");
const hooks_1 = require("@feathersjs/hooks");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const axios_1 = tslib_1.__importDefault(require("axios"));
const fs_extra_1 = tslib_1.__importDefault(require("fs-extra"));
const path = tslib_1.__importStar(require("path"));
const globalVars_1 = require("../common/globalVars");
const localizeUtils_1 = require("../common/localizeUtils");
const actionInjector_1 = require("../component/configManager/actionInjector");
const constants_1 = require("../component/constants");
const questionMW_1 = require("../component/middleware/questionMW");
const pathUtils_1 = require("../component/utils/pathUtils");
const constants_2 = require("../question/constants");
const concurrentLocker_1 = require("./middleware/concurrentLocker");
const errorHandler_1 = require("./middleware/errorHandler");
class FxCoreDeclarativeAgentPart {
    async updateActionWithMCP(inputs) {
        var _a, _b, _c, _d;
        const context = (0, globalVars_1.createContext)();
        const projectPath = inputs.projectPath;
        if (!projectPath) {
            throw new Error("projectPath is undefined"); // should never happen
        }
        const aiPluginFilePath = inputs[constants_2.QuestionNames.PluginManifestFilePath];
        if (!(await fs_extra_1.default.pathExists(aiPluginFilePath))) {
            const error = new teamsfx_api_1.SystemError("MCPForDAPluginManifestNotFound", "PluginManifestNotFound", (0, localizeUtils_1.getDefaultString)("core.MCPForDA.pluginManifestNotFound", aiPluginFilePath), (0, localizeUtils_1.getLocalizedString)("core.MCPForDA.pluginManifestNotFound", aiPluginFilePath));
            return (0, teamsfx_api_1.err)(error);
        }
        const aiPluginFilePathRelative = path.basename(aiPluginFilePath);
        const mcpServerUrl = inputs[constants_2.QuestionNames.MCPForDAServerUrl];
        const serverName = inputs[constants_2.QuestionNames.MCPForDAServerName];
        const mcpAuth = inputs[constants_2.QuestionNames.MCPForDAAuth];
        const authType = inputs[constants_2.QuestionNames.MCPForDAAuthType];
        let oauthAuthorizationUrl = undefined;
        let oauthTokenUrl = undefined;
        let oauthRefreshUrl = undefined;
        let registrationId = undefined;
        if (mcpAuth === "OAuthPluginVault") {
            try {
                registrationId = `MCP_DA_AUTH_ID_${serverName.toUpperCase()}`;
                if (authType === "oauth") {
                    let wellKnownMetadataUrl = inputs[constants_2.QuestionNames.MCPForDAAuthWellKnownUrl];
                    if (!wellKnownMetadataUrl) {
                        const mcpAuthMetadataUrl = inputs[constants_2.QuestionNames.MCPForDAAuthMetadataUrl];
                        if (!mcpAuthMetadataUrl) {
                            throw new Error((0, localizeUtils_1.getLocalizedString)("core.MCPForDA.mcpAuthMetadataUrlNotFound"));
                        }
                        const response = await axios_1.default.get(mcpAuthMetadataUrl);
                        if (response.status === 200 &&
                            response.data &&
                            response.data.authorization_servers &&
                            response.data.authorization_servers.length > 0) {
                            const mcpServerMetadataUrl = (_a = response.data.authorization_servers) === null || _a === void 0 ? void 0 : _a[0];
                            // Transform the URL to the proper OAuth authorization server metadata endpoint
                            // According to RFC 8414, the well-known endpoint should be constructed as:
                            // https://{domain}/.well-known/oauth-authorization-server{path}
                            const serverUrl = new URL(mcpServerMetadataUrl);
                            wellKnownMetadataUrl = `${serverUrl.protocol}//${serverUrl.host}/.well-known/oauth-authorization-server${serverUrl.pathname}`;
                        }
                        else {
                            throw new Error((0, localizeUtils_1.getLocalizedString)("core.MCPForDA.mcpServerMetadataUrlNotFound"));
                        }
                    }
                    const metadataResponse = await axios_1.default.get(wellKnownMetadataUrl);
                    if (metadataResponse.status === 200) {
                        oauthAuthorizationUrl = metadataResponse.data.authorization_endpoint;
                        oauthTokenUrl = metadataResponse.data.token_endpoint;
                        oauthRefreshUrl = metadataResponse.data.refresh_endpoint;
                    }
                    if (!oauthAuthorizationUrl || !oauthTokenUrl) {
                        throw new Error((0, localizeUtils_1.getLocalizedString)("core.MCPForDA.authUrlNotFound"));
                    }
                }
            }
            catch (error) {
                void context.userInteraction.showMessage("error", (0, localizeUtils_1.getLocalizedString)("core.MCPForDA.mcpAuthMetadataMissingError", error.message), false);
            }
        }
        // 2. Read ai-plugin.json
        const aiPluginContent = await fs_extra_1.default.readJSON(aiPluginFilePath);
        // For dynamic fetch tools, keep the functions empty and add runtime info
        // TODO: support dynamic fetch tools in the future
        const mcpToolsDetail = inputs[constants_2.QuestionNames.MCPForDAAvailableTools];
        const mcpToolsSelected = inputs[constants_2.QuestionNames.MCPForDAPreFetchTools];
        if (!mcpToolsDetail || !mcpToolsSelected) {
            const error = new teamsfx_api_1.UserError("MCPForDAPreFetchToolsNotFound", "PreFetchToolsNotFound", (0, localizeUtils_1.getDefaultString)("core.MCPForDA.preFetchToolsNotFound"), (0, localizeUtils_1.getLocalizedString)("core.MCPForDA.preFetchToolsNotFound"));
            return (0, teamsfx_api_1.err)(error);
        }
        const toolsSelectedPrevious = [];
        aiPluginContent.runtimes
            .filter((runtime) => (runtime.type === "RemoteMCPServer" && runtime.spec.url === mcpServerUrl) ||
            runtime.type === "LocalPlugin")
            .forEach((runtime) => {
            toolsSelectedPrevious.push(...runtime.run_for_functions);
        });
        aiPluginContent.functions = aiPluginContent.functions.filter((func) => !toolsSelectedPrevious.includes(func.name));
        aiPluginContent.functions = [
            ...aiPluginContent.functions,
            ...mcpToolsDetail
                .filter((tool) => mcpToolsSelected.includes(tool.name))
                .map((tool) => {
                if (inputs[constants_2.QuestionNames.MCPLocalServerIdentifier] != null) {
                    return {
                        name: tool.name,
                        description: tool.description,
                        parameters: {
                            type: tool.inputSchema.type || "object",
                            properties: tool.inputSchema.properties,
                            required: tool.inputSchema.required || [],
                        },
                    };
                }
                else {
                    return {
                        name: tool.name,
                        description: tool.description,
                    };
                }
            }),
        ];
        const matchedRuntime = aiPluginContent.runtimes.find((runtime) => runtime.type === "RemoteMCPServer" && runtime.spec.url === mcpServerUrl);
        aiPluginContent.runtimes = aiPluginContent.runtimes.filter((runtime) => (runtime.type !== "RemoteMCPServer" && runtime.type !== "LocalPlugin") ||
            runtime.spec.url !== mcpServerUrl);
        if (inputs[constants_2.QuestionNames.MCPLocalServerIdentifier] != null) {
            aiPluginContent.runtimes.push({
                type: "LocalPlugin",
                spec: {
                    local_endpoint: `${constants_1.LocalMcpPrefix}${inputs[constants_2.QuestionNames.MCPLocalServerIdentifier]}`,
                },
                run_for_functions: mcpToolsSelected,
            });
        }
        else {
            let mcpFile = (_b = matchedRuntime === null || matchedRuntime === void 0 ? void 0 : matchedRuntime.spec.mcp_tool_description) === null || _b === void 0 ? void 0 : _b.file;
            if (!mcpFile) {
                mcpFile = "mcp-tools.json";
                let suffix = 1;
                while (await fs_extra_1.default.pathExists(path.join(path.dirname(aiPluginFilePath), mcpFile))) {
                    mcpFile = `mcp-tools-${suffix}.json`;
                    suffix += 1;
                }
            }
            await fs_extra_1.default.writeJSON(path.join(path.dirname(aiPluginFilePath), mcpFile), {
                tools: [
                    ...mcpToolsDetail
                        .filter((tool) => mcpToolsSelected.includes(tool.name))
                        .map((tool) => {
                        return Object.assign(Object.assign({}, tool), { title: tool.name
                                .replace(/_/g, " ")
                                .replace(/^./, (str) => str.toUpperCase()) });
                    }),
                ],
            }, { spaces: 4 });
            aiPluginContent.runtimes.push({
                type: "RemoteMCPServer",
                spec: {
                    url: mcpServerUrl,
                    mcp_tool_description: {
                        file: mcpFile,
                    },
                },
                run_for_functions: mcpToolsSelected,
                auth: mcpAuth === "OAuthPluginVault" && !!registrationId
                    ? {
                        type: "OAuthPluginVault",
                        reference_id: `$\{\{${registrationId}\}\}`,
                    }
                    : {
                        type: "None",
                    },
            });
        }
        if (mcpAuth === "OAuthPluginVault" && !!registrationId) {
            // insert oauth info in teamsapp.yaml
            await actionInjector_1.ActionInjector.injectCreateOAuthActionForMCP(pathUtils_1.pathUtils.getYmlFilePath(projectPath), authType, serverName, registrationId, mcpServerUrl, oauthAuthorizationUrl, oauthTokenUrl, oauthRefreshUrl);
        }
        void context.userInteraction
            .showMessage("info", (0, localizeUtils_1.getLocalizedString)("core.MCPForDA.updatePluginManifest", aiPluginFilePathRelative), false, "Provision")
            .then((result) => {
            if (result.isOk() && result.value === "Provision") {
                void this.provisionResources(inputs);
            }
        });
        await fs_extra_1.default.writeJSON(aiPluginFilePath, aiPluginContent, { spaces: 4 });
        void ((_d = (_c = context.userInteraction).openFile) === null || _d === void 0 ? void 0 : _d.call(_c, aiPluginFilePath));
        return (0, teamsfx_api_1.ok)(undefined);
    }
    // This method will be implemented by FxCore
    provisionResources(_inputs) {
        throw new Error("Method not implemented.");
    }
}
tslib_1.__decorate([
    (0, hooks_1.hooks)([
        (0, globalVars_1.ErrorContextMW)({ component: "FxCore", stage: teamsfx_api_1.Stage.installApp }),
        errorHandler_1.ErrorHandlerMW,
        (0, questionMW_1.QuestionMW)("updateActionWithMCP"),
        concurrentLocker_1.ConcurrentLockerMW,
    ]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", Promise)
], FxCoreDeclarativeAgentPart.prototype, "updateActionWithMCP", null);
exports.FxCoreDeclarativeAgentPart = FxCoreDeclarativeAgentPart;
