// @microsoft/teamsfx-core/build/component/utils/odrProvider.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ODRProvider = void 0;
const child_process_1 = require("child_process");
const util_1 = require("util");
class ODRProvider {
    /**
     * Check if a server configuration is ODR-based.
     * @param serverConfig The server configuration object
     * @returns True if the server is ODR-based, false otherwise
     */
    static isODRServer(serverConfig) {
        if ((serverConfig === null || serverConfig === void 0 ? void 0 : serverConfig.type) !== "stdio" || !serverConfig.command) {
            return false;
        }
        const configCommand = serverConfig.command.toLowerCase();
        return configCommand === "odr" || configCommand.endsWith("odr.exe");
    }
    /**
     * Parse the output of 'odr list' command
     * @param jsonOutput The JSON output from 'odr list' command
     * @returns Array of parsed ODR servers with their tools
     */
    static parseODRListOutput(jsonOutput) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        const servers = [];
        if (!jsonOutput || !jsonOutput.servers || !Array.isArray(jsonOutput.servers)) {
            return servers;
        }
        for (const server of jsonOutput.servers) {
            const manifest = (_c = (_b = (_a = server._meta) === null || _a === void 0 ? void 0 : _a["io.modelcontextprotocol.registry/publisher-provided"]) === null || _b === void 0 ? void 0 : _b["com.microsoft.windows"]) === null || _c === void 0 ? void 0 : _c.manifest;
            const staticResponses = (_e = (_d = manifest === null || manifest === void 0 ? void 0 : manifest._meta) === null || _d === void 0 ? void 0 : _d["com.microsoft.windows"]) === null || _e === void 0 ? void 0 : _e.static_responses;
            const toolsList = ((_f = staticResponses === null || staticResponses === void 0 ? void 0 : staticResponses["tools/list"]) === null || _f === void 0 ? void 0 : _f.tools) || [];
            const packageFamily = (_h = (_g = manifest === null || manifest === void 0 ? void 0 : manifest._meta) === null || _g === void 0 ? void 0 : _g["com.microsoft.windows"]) === null || _h === void 0 ? void 0 : _h.package_family_name;
            const mcpConfig = (_j = manifest === null || manifest === void 0 ? void 0 : manifest.server) === null || _j === void 0 ? void 0 : _j.mcp_config;
            if (packageFamily && mcpConfig) {
                servers.push({
                    name: server.name,
                    packageFamily: packageFamily,
                    display_name: (manifest === null || manifest === void 0 ? void 0 : manifest.display_name) || server.name,
                    description: server.description || "",
                    version: server.version || "1.0.0",
                    identifier: ((_l = (_k = server.packages) === null || _k === void 0 ? void 0 : _k[0]) === null || _l === void 0 ? void 0 : _l.identifier) || "",
                    command: mcpConfig.command || "",
                    args: mcpConfig.args || [],
                    tools: toolsList.map((tool) => ({
                        name: tool.name,
                        description: tool.description || "",
                        inputSchema: tool.inputSchema,
                        outputSchema: tool.outputSchema,
                    })),
                });
            }
        }
        return servers;
    }
    /**
     * List all available local MCP servers from Windows ODR.
     * This is used during project creation to show available servers to the user.
     * @returns Array of ODR servers. Returns empty array if:
     *   - Not on Windows platform
     *   - ODR command is not available (not installed)
     *   - ODR command fails or returns invalid output
     */
    static async listServers() {
        if (process.platform !== "win32") {
            return [];
        }
        const execAsync = (0, util_1.promisify)(child_process_1.exec);
        try {
            const { stdout } = await execAsync("odr list");
            if (!stdout) {
                return [];
            }
            const jsonOutput = JSON.parse(stdout);
            return ODRProvider.parseODRListOutput(jsonOutput);
        }
        catch (error) {
            console.error("Error executing odr list:", error);
            return [];
        }
    }
    /**
     * Get tools for a specific ODR server by matching command and arguments.
     * @param command The command of the ODR server (e.g., "odr")
     * @param args The arguments array for the ODR server
     * @returns Array of tools from the matching ODR server, or empty array if no match found
     */
    static async getToolsForODRServer(command, args = []) {
        const odrServers = await ODRProvider.listServers();
        const matchingServer = odrServers.find((odrServer) => {
            return (odrServer.command === command && JSON.stringify(odrServer.args) === JSON.stringify(args));
        });
        return (matchingServer === null || matchingServer === void 0 ? void 0 : matchingServer.tools) || [];
    }
}
exports.ODRProvider = ODRProvider;
