// @microsoft/teamsfx-core/build/common/featureFlags.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.featureFlagManager = exports.FeatureFlagManager = exports.FeatureFlags = exports.FeatureFlagName = exports.isFeatureFlagEnabled = void 0;
// Determine whether feature flag is enabled based on environment variable setting
function isFeatureFlagEnabled(featureFlagName, defaultValue = false) {
    const flag = process.env[featureFlagName];
    if (flag === undefined) {
        return defaultValue; // allows consumer to set a default value when environment variable not set
    }
    else {
        return flag === "1" || flag.toLowerCase() === "true"; // can enable feature flag by set environment variable value to "1" or "true"
    }
}
exports.isFeatureFlagEnabled = isFeatureFlagEnabled;
class FeatureFlagName {
}
exports.FeatureFlagName = FeatureFlagName;
FeatureFlagName.CLIDotNet = "TEAMSFX_CLI_DOTNET";
FeatureFlagName.SampleConfigBranch = "TEAMSFX_SAMPLE_CONFIG_BRANCH";
FeatureFlagName.ChatParticipantUIEntries = "TEAMSFX_CHAT_PARTICIPANT_ENTRIES";
FeatureFlagName.HideGitHubCopilotPreviewTag = "TEAMSFX_HIDE_GITHUB_COPILOT_PREVIEW_TAG";
FeatureFlagName.SMEOAuth = "SME_OAUTH";
FeatureFlagName.ShowDiagnostics = "TEAMSFX_SHOW_DIAGNOSTICS";
FeatureFlagName.TelemetryTest = "TEAMSFX_TELEMETRY_TEST";
FeatureFlagName.DevTunnelTest = "TEAMSFX_DEV_TUNNEL_TEST";
FeatureFlagName.SyncManifest = "TEAMSFX_SYNC_MANIFEST";
FeatureFlagName.KiotaNPMIntegration = "TEAMSFX_KIOTA_NPM_INTEGRATION";
FeatureFlagName.CEAEnabled = "TEAMSFX_CEA_ENABLED";
FeatureFlagName.EmbeddedKnowledgeEnabled = "TEAMSFX_EMBEDDED_KNOWLEDGE";
FeatureFlagName.SandBoxedTeam = "TEAMSFX_SANDBOXED_TEAM";
FeatureFlagName.SensitivityLabelEnabled = "TEAMSFX_SENSITIVITY_LABEL";
FeatureFlagName.DAMetaOS = "TEAMSFX_DA_METAOS";
FeatureFlagName.CFShortcutMetaOS = "TEAMSFX_CF_SHORTCUT_METAOS";
FeatureFlagName.MCPForDA = "TEAMSFX_MCP_FOR_DA";
FeatureFlagName.BrokerAuth = "TEAMSFX_BROKER_AUTH";
// Add config files to existing project to make it toolkit compatible
FeatureFlagName.GenerateConfigFiles = "TEAMSFX_GENERATE_CONFIG_FILES";
// Permanent feature flag for sovereign cloud environment setting
FeatureFlagName.SovereignCloudEnvironment = "TEAMSFX_SOVEREIGN_CLOUD_ENVIRONMENT";
class FeatureFlags {
}
exports.FeatureFlags = FeatureFlags;
// Testing feature flags - not for production use
FeatureFlags.CLIDotNet = { name: FeatureFlagName.CLIDotNet, defaultValue: "false" };
FeatureFlags.ChatParticipantUIEntries = {
    name: FeatureFlagName.ChatParticipantUIEntries,
    defaultValue: "false",
};
FeatureFlags.HideGitHubCopilotPreviewTag = {
    name: FeatureFlagName.HideGitHubCopilotPreviewTag,
    defaultValue: "false",
};
FeatureFlags.SMEOAuth = { name: FeatureFlagName.SMEOAuth, defaultValue: "false" };
FeatureFlags.ShowDiagnostics = {
    name: FeatureFlagName.ShowDiagnostics,
    defaultValue: "false",
};
FeatureFlags.TelemetryTest = {
    name: FeatureFlagName.TelemetryTest,
    defaultValue: "false",
};
FeatureFlags.DevTunnelTest = {
    name: FeatureFlagName.DevTunnelTest,
    defaultValue: "false",
};
FeatureFlags.SyncManifest = {
    name: FeatureFlagName.SyncManifest,
    defaultValue: "false",
};
FeatureFlags.KiotaNPMIntegration = {
    name: FeatureFlagName.KiotaNPMIntegration,
    defaultValue: "true",
};
FeatureFlags.CEAEnabled = {
    name: FeatureFlagName.CEAEnabled,
    defaultValue: "false",
};
FeatureFlags.EmbeddedKnowledgeEnabled = {
    name: FeatureFlagName.EmbeddedKnowledgeEnabled,
    defaultValue: "true",
};
FeatureFlags.SandBoxedTeam = {
    name: FeatureFlagName.SandBoxedTeam,
    defaultValue: "false",
};
FeatureFlags.SensitivityLabelEnabled = {
    name: FeatureFlagName.SensitivityLabelEnabled,
    defaultValue: "false",
};
FeatureFlags.DAMetaOS = {
    name: FeatureFlagName.DAMetaOS,
    defaultValue: "false",
};
FeatureFlags.CFShortcutMetaOS = {
    name: FeatureFlagName.CFShortcutMetaOS,
    defaultValue: "false",
};
FeatureFlags.MCPForDA = {
    name: FeatureFlagName.MCPForDA,
    defaultValue: "true",
};
FeatureFlags.BrokerAuth = {
    name: FeatureFlagName.BrokerAuth,
    defaultValue: "false",
};
FeatureFlags.GenerateConfigFiles = {
    name: FeatureFlagName.GenerateConfigFiles,
    defaultValue: "false",
};
FeatureFlags.SovereignCloudEnvironment = {
    name: FeatureFlagName.SovereignCloudEnvironment,
    defaultValue: "",
};
class FeatureFlagManager {
    getBooleanValue(featureFlag) {
        return isFeatureFlagEnabled(featureFlag.name, featureFlag.defaultValue === "true" || featureFlag.defaultValue === "1");
    }
    setBooleanValue(featureFlag, value) {
        process.env[featureFlag.name] = value ? "true" : "false";
    }
    getStringValue(featureFlag) {
        return process.env[featureFlag.name] || featureFlag.defaultValue;
    }
    list() {
        return Object.values(FeatureFlags);
    }
    listEnabled() {
        return this.list()
            .filter((f) => isFeatureFlagEnabled(f.name))
            .map((f) => f.name);
    }
}
exports.FeatureFlagManager = FeatureFlagManager;
exports.featureFlagManager = new FeatureFlagManager();
