// @microsoft/app-manifest/build/wrappers/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComposeExtensionTypeEnum = exports.ConfigurableTabScope = exports.StaticTabScope = exports.DefaultInstallScope = exports.TeamsManifestWrapper = exports.CapabilityName = exports.DeclarativeAgentManifestWrapper = exports.RuntimeType = exports.APIPluginManifestWrapper = exports.BaseManifest = void 0;
/**
 * OOP Wrappers for Manifest Types
 *
 * This module provides object-oriented wrappers around the generated manifest types,
 * offering a fluent API for manipulation, state tracking, and convenient operations.
 *
 * @module wrappers
 */
var BaseManifest_1 = require("./BaseManifest");
Object.defineProperty(exports, "BaseManifest", { enumerable: true, get: function () { return BaseManifest_1.BaseManifest; } });
var APIPluginManifestWrapper_1 = require("./APIPluginManifestWrapper");
Object.defineProperty(exports, "APIPluginManifestWrapper", { enumerable: true, get: function () { return APIPluginManifestWrapper_1.APIPluginManifestWrapper; } });
Object.defineProperty(exports, "RuntimeType", { enumerable: true, get: function () { return APIPluginManifestWrapper_1.RuntimeType; } });
var DeclarativeAgentManifestWrapper_1 = require("./DeclarativeAgentManifestWrapper");
Object.defineProperty(exports, "DeclarativeAgentManifestWrapper", { enumerable: true, get: function () { return DeclarativeAgentManifestWrapper_1.DeclarativeAgentManifestWrapper; } });
Object.defineProperty(exports, "CapabilityName", { enumerable: true, get: function () { return DeclarativeAgentManifestWrapper_1.CapabilityName; } });
var TeamsManifestWrapper_1 = require("./TeamsManifestWrapper");
Object.defineProperty(exports, "TeamsManifestWrapper", { enumerable: true, get: function () { return TeamsManifestWrapper_1.TeamsManifestWrapper; } });
Object.defineProperty(exports, "DefaultInstallScope", { enumerable: true, get: function () { return TeamsManifestWrapper_1.DefaultInstallScope; } });
Object.defineProperty(exports, "StaticTabScope", { enumerable: true, get: function () { return TeamsManifestWrapper_1.StaticTabScope; } });
Object.defineProperty(exports, "ConfigurableTabScope", { enumerable: true, get: function () { return TeamsManifestWrapper_1.ConfigurableTabScope; } });
Object.defineProperty(exports, "ComposeExtensionTypeEnum", { enumerable: true, get: function () { return TeamsManifestWrapper_1.ComposeExtensionTypeEnum; } });
