// @microsoft/teamsfx-core/build/question/scaffold/vsc/teamsProjectTypeNode.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateActionWithMCP = exports.MCPForDAServerUrlNode = exports.MCPLocalServerSelectionNode = exports.MCPServerTypeNode = exports.m365SearchMeSubNode = exports.meProjectTypeNode = exports.tabProjectTypeNode = exports.botProjectTypeNode = exports.notificationBotTriggerNode = exports.customCopilotRagNode = exports.TeamsProjectTypeOptions = exports.getTeamsProjectNode = void 0;
const tslib_1 = require("tslib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const fs = tslib_1.__importStar(require("fs-extra"));
const os_1 = tslib_1.__importDefault(require("os"));
const path_1 = tslib_1.__importDefault(require("path"));
const localizeUtils_1 = require("../../../common/localizeUtils");
const templateHelper_1 = require("../../../component/generator/templateHelper");
const odrProvider_1 = require("../../../component/utils/odrProvider");
const folder_1 = require("../../../folder");
const create_1 = require("../../create");
const questionNames_1 = require("../../questionNames");
const commonNodes_1 = require("../commonNodes");
const constructNode_1 = require("../constructNode");
const CapabilityOptions_1 = require("./CapabilityOptions");
function getTeamsProjectNode() {
    let jsonPath;
    const cachedJsonPath = path_1.default.join(os_1.default.homedir(), `.${String(teamsfx_api_1.ConfigFolderName)}`, "ui", "teamsNode.json");
    // Check if cached JSON exists, otherwise fallback to bundled templates folder
    if (!(0, templateHelper_1.useLocalTemplate)() && fs.pathExistsSync(cachedJsonPath)) {
        jsonPath = cachedJsonPath;
    }
    else {
        jsonPath = path_1.default.join((0, folder_1.getTemplatesFolder)(), "ui", "teamsNode.json");
    }
    const content = fs.readFileSync(jsonPath, "utf-8");
    return (0, constructNode_1.constructNode)(content);
}
exports.getTeamsProjectNode = getTeamsProjectNode;
class TeamsProjectTypeOptions {
    static tab(platform = teamsfx_api_1.Platform.VSCode) {
        return {
            id: TeamsProjectTypeOptions.tabOptionId,
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(browser) " : ""}${(0, localizeUtils_1.getLocalizedString)("core.TabOption.label")}`,
            detail: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.tab.detail"),
        };
    }
    static bot(platform = teamsfx_api_1.Platform.VSCode) {
        return {
            id: TeamsProjectTypeOptions.botOptionId,
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(hubot) " : ""}${(0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.bot.label")}`,
            detail: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.bot.detail"),
        };
    }
    static me(platform = teamsfx_api_1.Platform.VSCode) {
        return {
            id: TeamsProjectTypeOptions.meOptionId,
            label: `${platform === teamsfx_api_1.Platform.VSCode ? "$(symbol-keyword) " : ""}${(0, localizeUtils_1.getLocalizedString)("core.MessageExtensionOption.label")}`,
            detail: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.messageExtension.copilotEnabled.detail"),
        };
    }
}
exports.TeamsProjectTypeOptions = TeamsProjectTypeOptions;
TeamsProjectTypeOptions.tabOptionId = "tab-type";
TeamsProjectTypeOptions.botOptionId = "bot-type";
TeamsProjectTypeOptions.meOptionId = "me-type";
function customCopilotRagNode() {
    return {
        condition: { equals: CapabilityOptions_1.TeamsAgentCapabilityOptions.customCopilotRag().id },
        data: {
            type: "singleSelect",
            name: questionNames_1.QuestionNames.CustomCopilotRag,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.capability.customCopilotRagOption.label"),
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.capability.customCopilotRag.placeholder"),
            staticOptions: [
                CapabilityOptions_1.CustomCopilotRagOptions.customize(),
                CapabilityOptions_1.CustomCopilotRagOptions.azureAISearch(),
                CapabilityOptions_1.CustomCopilotRagOptions.customApi(),
                // CustomCopilotRagOptions.microsoft365(),
            ],
            default: CapabilityOptions_1.CustomCopilotRagOptions.customize().id,
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
        children: [(0, commonNodes_1.apiSpecNode)({ equals: CapabilityOptions_1.CustomCopilotRagOptions.customApi().id })],
    };
}
exports.customCopilotRagNode = customCopilotRagNode;
// export function aiAgentNode(): IQTreeNode {
//   return {
//     condition: { equals: CustomCopilotCapabilityOptions.aiAgent().id },
//     data: {
//       type: "singleSelect",
//       name: QuestionNames.CustomCopilotAssistant,
//       title: getLocalizedString(
//         "core.createProjectQuestion.capability.customCopilotAssistant.title"
//       ),
//       placeholder: getLocalizedString(
//         "core.createProjectQuestion.capability.customCopilotAssistant.placeholder"
//       ),
//       staticOptions: [
//         CustomCopilotAssistantOptions.new(),
//         CustomCopilotAssistantOptions.assistantsApi(),
//       ],
//       default: CustomCopilotAssistantOptions.new().id,
//       onDidSelection: setTemplateName,
//     },
//   };
// }
function notificationBotTriggerNode(platform = teamsfx_api_1.Platform.VSCode) {
    return {
        condition: { equals: CapabilityOptions_1.BotCapabilityOptions.notificationBotId },
        data: {
            name: questionNames_1.QuestionNames.BotTrigger,
            title: (0, localizeUtils_1.getLocalizedString)("plugins.bot.questionHostTypeTrigger.title"),
            type: "singleSelect",
            cliDescription: "Specifies the trigger for `Chat Notification Message` app template.",
            staticOptions: [
                platform === teamsfx_api_1.Platform.VS
                    ? CapabilityOptions_1.NotificationBotOptions.appServiceForVS()
                    : CapabilityOptions_1.NotificationBotOptions.appService(),
                CapabilityOptions_1.NotificationBotOptions.functionsHttpAndTimerTrigger(),
                CapabilityOptions_1.NotificationBotOptions.functionsHttpTrigger(),
                CapabilityOptions_1.NotificationBotOptions.functionsTimerTrigger(),
            ],
            default: platform === teamsfx_api_1.Platform.VS
                ? CapabilityOptions_1.NotificationBotOptions.appServiceForVS().id
                : CapabilityOptions_1.NotificationBotOptions.appService().id,
            placeholder: (0, localizeUtils_1.getLocalizedString)("plugins.bot.questionHostTypeTrigger.placeholder"),
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
    };
}
exports.notificationBotTriggerNode = notificationBotTriggerNode;
function botProjectTypeNode() {
    return {
        // project-type = Bot
        condition: { equals: TeamsProjectTypeOptions.botOptionId },
        data: {
            name: questionNames_1.QuestionNames.Capabilities,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.bot.title"),
            type: "singleSelect",
            staticOptions: [
                CapabilityOptions_1.BotCapabilityOptions.basicBot(),
                CapabilityOptions_1.BotCapabilityOptions.notificationBot(),
                CapabilityOptions_1.BotCapabilityOptions.commandBot(),
                CapabilityOptions_1.BotCapabilityOptions.workflowBot(),
            ],
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createCapabilityQuestion.placeholder"),
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
        children: [notificationBotTriggerNode()],
    };
}
exports.botProjectTypeNode = botProjectTypeNode;
function tabProjectTypeNode(platform = teamsfx_api_1.Platform.VSCode) {
    return {
        // project-type = Tab
        condition: { equals: TeamsProjectTypeOptions.tabOptionId },
        data: {
            name: questionNames_1.QuestionNames.Capabilities,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.tab.title"),
            type: "singleSelect",
            staticOptions: [
                CapabilityOptions_1.TabCapabilityOptions.nonSsoTab(),
                CapabilityOptions_1.TabCapabilityOptions.m365SsoLaunchPage(),
                CapabilityOptions_1.TabCapabilityOptions.dashboardTab(),
                CapabilityOptions_1.TabCapabilityOptions.SPFxTab(),
            ],
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createCapabilityQuestion.placeholder"),
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
        children: [
            {
                //SPFx sub-tree
                condition: { equals: CapabilityOptions_1.TabCapabilityOptions.SPFxTab().id },
                data: (0, create_1.SPFxSolutionQuestion)(),
                children: [
                    {
                        data: { type: "group" },
                        children: [
                            { data: (0, create_1.SPFxPackageSelectQuestion)() },
                            { data: (0, create_1.SPFxFrameworkQuestion)() },
                            { data: (0, create_1.SPFxWebpartNameQuestion)() },
                        ],
                        condition: { equals: "new" },
                    },
                    {
                        data: (0, create_1.SPFxImportFolderQuestion)(),
                        condition: { equals: "import" },
                    },
                ],
            },
        ],
    };
}
exports.tabProjectTypeNode = tabProjectTypeNode;
function meProjectTypeNode() {
    return {
        // project-type = Messaging Extension
        condition: { equals: TeamsProjectTypeOptions.meOptionId },
        data: {
            name: questionNames_1.QuestionNames.Capabilities,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.messageExtension.title"),
            type: "singleSelect",
            staticOptions: [
                CapabilityOptions_1.MeCapabilityOptions.m365SearchMe(),
                CapabilityOptions_1.MeCapabilityOptions.collectFormMe(),
                CapabilityOptions_1.MeCapabilityOptions.linkUnfurling(),
            ],
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createCapabilityQuestion.placeholder"),
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
        children: [m365SearchMeSubNode()],
    };
}
exports.meProjectTypeNode = meProjectTypeNode;
function m365SearchMeSubNode() {
    return {
        // Search ME sub-tree
        condition: { equals: CapabilityOptions_1.MeCapabilityOptions.m365SearchMe().id },
        data: {
            name: questionNames_1.QuestionNames.MeArchitectureType,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.meArchitecture.title"),
            cliDescription: "The authentication type for the API.",
            type: "singleSelect",
            staticOptions: [
                CapabilityOptions_1.MeArchitectureOptions.newApi(),
                CapabilityOptions_1.MeArchitectureOptions.openApiSpec(),
                CapabilityOptions_1.MeArchitectureOptions.botMe(),
            ],
            default: CapabilityOptions_1.MeArchitectureOptions.newApi().id,
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.projectType.copilotExtension.placeholder"),
            forgetLastValue: true,
            skipSingleOption: true,
            onDidSelection: CapabilityOptions_1.setTemplateName,
        },
        children: [
            {
                condition: { equals: CapabilityOptions_1.MeArchitectureOptions.newApi().id },
                data: {
                    type: "singleSelect",
                    name: questionNames_1.QuestionNames.ApiAuth,
                    title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.apiMessageExtensionAuth.title"),
                    placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.apiMessageExtensionAuth.placeholder"),
                    staticOptions: [
                        CapabilityOptions_1.ApiAuthOptions.none(true),
                        CapabilityOptions_1.ApiAuthOptions.bearerToken(),
                        CapabilityOptions_1.ApiAuthOptions.microsoftEntra(true),
                    ],
                    default: CapabilityOptions_1.ApiAuthOptions.none(true).id,
                    onDidSelection: CapabilityOptions_1.setTemplateName,
                },
            },
            (0, commonNodes_1.apiSpecNode)({ equals: CapabilityOptions_1.MeArchitectureOptions.openApiSpec().id }),
        ],
    };
}
exports.m365SearchMeSubNode = m365SearchMeSubNode;
function MCPServerTypeNode() {
    return {
        condition: { equals: CapabilityOptions_1.ActionStartOptions.mcp().id },
        data: {
            name: questionNames_1.QuestionNames.MCPServerType,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.title"),
            type: "singleSelect",
            staticOptions: [],
            dynamicOptions: async (inputs) => {
                const servers = await odrProvider_1.ODRProvider.listServers();
                inputs["_McpOdrOutput"] = servers;
                const options = [
                    {
                        id: "remote",
                        label: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.remote.label"),
                        detail: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.remote.detail"),
                    },
                ];
                if (servers.length > 0) {
                    options.push({
                        id: "local",
                        label: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.local.label"),
                        detail: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.local.detail"),
                    });
                }
                return options;
            },
            default: "remote",
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpServerType.placeholder"),
            skipSingleOption: true,
        },
        children: [
            {
                condition: { equals: "remote" },
                data: MCPForDAServerUrlNode().data,
            },
            MCPLocalServerSelectionNode(),
        ],
    };
}
exports.MCPServerTypeNode = MCPServerTypeNode;
function MCPLocalServerSelectionNode() {
    return {
        condition: { equals: "local" },
        data: {
            name: questionNames_1.QuestionNames.MCPLocalServer,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpLocalServer.title"),
            type: "multiSelect",
            returnObject: true,
            staticOptions: [],
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpLocalServer.placeholder"),
            dynamicOptions: (inputs) => {
                const servers = inputs["_McpOdrOutput"];
                return servers.map((server) => ({
                    id: server.name,
                    label: server.display_name,
                    detail: `${server.description} (${server.tools.length} tools available)`,
                    data: JSON.stringify({
                        identifier: server.identifier,
                        command: server.command,
                        args: server.args,
                    }),
                }));
            },
            validation: { minItems: 1 },
            validationHelp: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpLocalServer.validationHelp"),
        },
        children: [],
    };
}
exports.MCPLocalServerSelectionNode = MCPLocalServerSelectionNode;
function MCPForDAServerUrlNode() {
    return {
        condition: { equals: CapabilityOptions_1.ActionStartOptions.mcp().id },
        data: {
            name: questionNames_1.QuestionNames.MCPForDAServerUrl,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.ServerUrl.title"),
            type: "text",
            placeholder: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.ServerUrl.placeholder"),
        },
    };
}
exports.MCPForDAServerUrlNode = MCPForDAServerUrlNode;
function updateActionWithMCP() {
    return {
        data: {
            type: "singleFile",
            name: questionNames_1.QuestionNames.PluginManifestFilePath,
            title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.File.title"),
            defaultFolder: (inputs) => path_1.default.normalize(inputs.projectPath),
            default: (inputs) => path_1.default.normalize(path_1.default.join(inputs.projectPath, teamsfx_api_1.AppPackageFolderName, teamsfx_api_1.DefaultPluginManifestFileName)),
        },
        children: [
            {
                data: {
                    type: "multiSelect",
                    name: questionNames_1.QuestionNames.MCPForDAPreFetchTools,
                    title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.PreFetchTools.title"),
                    staticOptions: [],
                    dynamicOptions: (inputs) => {
                        const availableTools = inputs[questionNames_1.QuestionNames.MCPForDAAvailableTools];
                        const tools = availableTools.map((tool) => {
                            return {
                                id: tool.name,
                                label: tool.name,
                                detail: tool.description || "",
                            };
                        });
                        return tools;
                    },
                    default: async (inputs) => {
                        const pluginManifestFilePath = inputs[questionNames_1.QuestionNames.PluginManifestFilePath];
                        if (!pluginManifestFilePath) {
                            return [];
                        }
                        const pluginManifest = await fs.readJSON(pluginManifestFilePath);
                        const serverUrl = inputs[questionNames_1.QuestionNames.MCPForDAServerUrl];
                        const result = [];
                        pluginManifest.runtimes
                            .filter((runtime) => runtime.type === "RemoteMCPServer" &&
                            runtime.spec.url === serverUrl &&
                            !runtime.spec["enable_dynamic_discovery"])
                            .forEach((runtime) => {
                            result.push(...runtime["run_for_functions"]);
                        });
                        return result;
                    },
                },
            },
            {
                condition: (inputs) => {
                    return inputs[questionNames_1.QuestionNames.MCPForDAAuth] !== "NoneAuth";
                },
                data: {
                    type: "singleSelect",
                    name: questionNames_1.QuestionNames.MCPForDAAuthType,
                    title: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.AuthType.title"),
                    staticOptions: [
                        {
                            id: "oauth",
                            label: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.Auth.OAuth"),
                        },
                        {
                            id: "entraSSO",
                            label: (0, localizeUtils_1.getLocalizedString)("core.createProjectQuestion.mcpForDa.Auth.EntraSSO"),
                        },
                    ],
                    default: "oauth",
                },
            },
        ],
    };
}
exports.updateActionWithMCP = updateActionWithMCP;
